include/aml.h

```diff
@@ -96,53 +181,48 @@ void aml_interrupt(struct aml*);
  *
  * Returns how many references there were BEFORE the call.
  */
-int aml_ref(void* obj);
+int aml_loop_ref(struct aml* loop);
+int aml_handler_ref(struct aml_handler* obj);
+int aml_timer_ref(struct aml_timer* obj);
+int aml_ticker_ref(struct aml_ticker* obj);
+int aml_signal_ref(struct aml_signal* obj);
+int aml_work_ref(struct aml_work* obj);
+int aml_idle_ref(struct aml_idle* obj);
 
 /* Decrement the reference count by one.
  *
  * Returns how many references there are AFTER the call.
  */
-int aml_unref(void* obj);
-
-/* Create a new weak reference to the object.
- *
- * The reference object must be deleted using aml_weak_ref_del().
- */
-struct aml_weak_ref* aml_weak_ref_new(void* obj);
-
-/* Delete a weak reference created by aml_weak_ref_new().
- */
-void aml_weak_ref_del(struct aml_weak_ref* self);
-
-/* Try to get a new strong reference from a weak reference object.
- *
- * If the weak reference is still valid, the reference count on the returned
- * aml object will be increased by one. Otherwise NULL is returned.
- */
-void* aml_weak_ref_read(struct aml_weak_ref* self);
+int aml_loop_unref(struct aml* loop);
+int aml_handler_unref(struct aml_handler* obj);
+int aml_timer_unref(struct aml_timer* obj);
+int aml_ticker_unref(struct aml_ticker* obj);
+int aml_signal_unref(struct aml_signal* obj);
+int aml_work_unref(struct aml_work* obj);
+int aml_idle_unref(struct aml_idle* obj);
 
 /* The following calls create event handler objects.
  *
  * An object will have a reference count of 1 upon creation and must be freed
  * using aml_unref().
  */
-struct aml_handler* aml_handler_new(int fd, aml_callback_fn, void* userdata,
-                                    aml_free_fn);
+struct aml_handler* aml_handler_new(int fd, void (*cb)(struct aml_handler*),
```

Comment:
Note, this is a breaking API change, because a callback taking a `void *` is (sadly) not compatible with one taking a struct pointer.

Reply:
Yes, but I've never declared the API to be stable for this, so that's fine. Besides, this library doesn't have many users.

Reply:
Cool, just wanted to double-check that it's known.

Reply:
I might declare it stable soon after these changes though.

I'm also considering whether I should remove the free_fn. It is a bit of a foot-gun, but it's not particularly bad because if you know not to use it unless you really know what you're doing, you can't shoot yourself in the foot with it by mistake. Whereas the type-safety stuff was possible to mess up by mistake even if you knew about it.

Reply:
I'd be in favor of dropping `free_fn` because I don't use it and don't think it's particularly useful, but yeah, it's not a big deal since I can just ignore it.